### ISSUE TITLE
vvenc does not build with C++17

### ISSUE THREAD

**Issue by uilianries** [fraunhoferhhi/vvenc#289]
Hi!

I tried to build the project by using `CMAKE_CXX_STANDARD 17` and it does not work. There is a deprecated feature related to `std::iterator`. Is there any change by support C++17 too as a superset?

#### Environment

OS: Mac OSX 13.4.1
Compiler: Clang 14.0.3
CPU: aarch64 (Mac M1)

#### Steps 

    sed -i '' 's/CMAKE_CXX_STANDARD 14/CMAKE_CXX_STANDARD 17/g' CMakeLists.txt
    cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=ON -DBUILD_SHARED_LIBS=ON
    cmake --build build/

#### Build log

```
% cmake --build build
...
cd /tmp/vvenc-1.8.0/build/source/Lib/vvenc && /Library/Developer/CommandLineTools/usr/bin/c++ -DMISSING_INTRIN_mm256_zeroupper -DVVENC_DYN_LINK -DVVENC_SOURCE -Dvvenc_EXPORTS -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../../../include -I/tmp/vvenc-1.8.0/build -I/tmp/vvenc-1.8.0/source/Lib/vvenc/. -I/tmp/vvenc-1.8.0/source/Lib/vvenc/.. -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../DecoderLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../EncoderLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib/x86 -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../apputils -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../../../thirdparty/nlohmann_json/single_include -isystem /tmp/vvenc-1.8.0/source/Lib/vvenc/../../../thirdparty -DTARGET_SIMD_X86 -DVVENC_ENABLE_THIRDPARTY_JSON -DVVENC_ENABLE_THIRDPARTY_JSON -O3 -DNDEBUG -std=gnu++17 -flto=thin -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Werror -Wno-deprecated-register -Wno-unused-const-variable -Wno-unknown-attributes -MD -MT source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/DepQuant.cpp.o -MF CMakeFiles/vvenc.dir/__/CommonLib/DepQuant.cpp.o.d -o CMakeFiles/vvenc.dir/__/CommonLib/DepQuant.cpp.o -c /tmp/vvenc-1.8.0/source/Lib/CommonLib/DepQuant.cpp
[ 41%] Building CXX object source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/InterPrediction.cpp.o
cd /tmp/vvenc-1.8.0/build/source/Lib/vvenc && /Library/Developer/CommandLineTools/usr/bin/c++ -DMISSING_INTRIN_mm256_zeroupper -DVVENC_DYN_LINK -DVVENC_SOURCE -Dvvenc_EXPORTS -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../../../include -I/tmp/vvenc-1.8.0/build -I/tmp/vvenc-1.8.0/source/Lib/vvenc/. -I/tmp/vvenc-1.8.0/source/Lib/vvenc/.. -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../DecoderLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../EncoderLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib/x86 -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../apputils -I/tmp/vvenc-1.8.0/source/Lib/vvenc/../../../thirdparty/nlohmann_json/single_include -isystem /tmp/vvenc-1.8.0/source/Lib/vvenc/../../../thirdparty -DTARGET_SIMD_X86 -DVVENC_ENABLE_THIRDPARTY_JSON -DVVENC_ENABLE_THIRDPARTY_JSON -O3 -DNDEBUG -std=gnu++17 -flto=thin -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Werror -Wno-deprecated-register -Wno-unused-const-variable -Wno-unknown-attributes -MD -MT source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/InterPrediction.cpp.o -MF CMakeFiles/vvenc.dir/__/CommonLib/InterPrediction.cpp.o.d -o CMakeFiles/vvenc.dir/__/CommonLib/InterPrediction.cpp.o -c /tmp/vvenc-1.8.0/source/Lib/CommonLib/InterPrediction.cpp
In file included from /tmp/vvenc-1.8.0/source/Lib/CommonLib/InterPrediction.cpp:48:
In file included from /tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib/InterPrediction.h:48:
In file included from /tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib/InterpolationFilter.h:51:
/tmp/vvenc-1.8.0/source/Lib/vvenc/../CommonLib/Unit.h:441:34: error: 'iterator<std::forward_iterator_tag, vvenc::TransformUnit>' is deprecated [-Werror,-Wdeprecated-declarations]
class UnitIterator : public std::iterator<std::forward_iterator_tag, T>
                                 ^
/tmp/vvenc-1.8.0/source/Lib/CommonLib/InterPrediction.cpp:1903:21: note: in instantiation of template class 'vvenc::UnitIterator<vvenc::TransformUnit>' requested here
  for (auto& currPU : CU::traverseTUs(cu))
                    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__iterator/iterator.h:24:29: note: 'iterator<std::forward_iterator_tag, vvenc::TransformUnit>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
1 error generated.
make[2]: *** [source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/InterPrediction.cpp.o] Error 1
make[1]: *** [source/Lib/vvenc/CMakeFiles/vvenc.dir/all] Error 2
make: *** [all] Error 2

```

#### Full build log

[build.log](https://github.com/fraunhoferhhi/vvenc/files/11957180/build.log)

**Comment by adamjw24** [comment]
C++ development committee, making life hard for people every day...

I guess this is possible to fix fairly easily by manually implementing the iterator traits: https://stackoverflow.com/questions/43268146/why-is-stditerator-deprecated

We have an internal pre-release code freeze, so this will probably only be addressed later, not in time for v1.9.0.

**Comment by uilianries** [comment]
@adamjw24 thank you for your quick answer!
